feat: AWS Cognito (#1051)
appsmithorg/appsmith-docs: website/docs/getting-started/setup/instance-configuration/authentication/openid-connect-oidc/aws-cognito.md

:::

## Configure a user pool

Log in to your AWS account, open AWS Cognito and follow the steps below:

1. Click the "Create User pool" button to create a new user pool.
2. **Step 1 - Configure sign-in experience**: select "Email" in the Cognito user pool sign-in options. Click *Next*

    ![](/img/Step1-AWS_Cognito.png)

3. **Step 2 - Configure security requirements:** select "No MFA" in Multi-factor authentication. Click *Next*. 

    ![](/img/Step2-AWS_Cognito.png)

4. **Step 3 - configure sign-up experience:** keep the default configurations and click *Next*.
5. **Step 4 -Configure message delivery:** you can choose either of the options - *Send email with Amazon SES* or *Send email with Cognito*. Please pick *Send email with Cognito* for this setup.
6.  **Step 5 - Integrate your app:**
    1. Enter a user pool name.
    2. Under *Initial app client*, add an app client name and select **Generate a client secret** under Client secret section.

    ![](/img/AWS-cognito_App_client.png)

7. **Step 6 - Review Configurations:** review all the selected configurations and proceed once done.

> 8. Click the user pool you've created, go to the integrations tab and follow the steps below 

Reviewer: From here onwards it should be a new section 

## Configure domain and app client

    1. In the domains section, click on the actions dropdown and select **Create a Cognito domain**. You can also select **Create a custom domain** to add a domain that you own.

    ![](/img/AWS-cognito_create-cognito-domain.png)

    2. Enter a domain prefix in the **Create Cognito domain** screen and click on **Create Cognito domain** button.

        ![](/img/AWS-cognito_create_domain.png)

> 9. On the user pool home, scroll down to the **App client list** and open the App Client that you created.

Reviewer: ```suggestion
9. On the **App Integration** tab, scroll down to the **App client list** section and click the name of the app client you created in the preceding steps.
```

>     1. Copy the client ID and client secret that will be added to the OIDC configurations in Appsmith.

Reviewer: ```suggestion
    1. Copy the **Client ID** and **Client Secret**.
```

    ![](/img/AWS-cognito_client-creds.png)

>     2. Scroll down to the Hosted UI section and click Edit. Click the Add Callback URL button and enter Redirect copied from the OIDC window in Appsmith’s Admin Settings.

Reviewer: This looks amiss as you have not yet talked about the Appsmith instance. meaning before creating the user pool, we should ask the user to enable the OIDC on Appsmuth and copy the redirect URL.

Author: If you look at the other OIDC provider docs, we have followed the same structure because the URLs are copied from both endpoints. Only Redirect Url goes from Appsmith to the OIDC provider, 3-4 URLs come from the OIDC provider to Appsmith. So we decided to explain the OIDC provider first and then discuss the steps in the Appsmith instance.

>     2. Scroll down to the Hosted UI section and click Edit. Click the Add Callback URL button and enter Redirect copied from the OIDC window in Appsmith’s Admin Settings.

Reviewer: ```suggestion
    2. Scroll down to the **Hosted UI** section and click **Edit**. Click the **Add Callback URL** button and enter **Redirect URL** copied from the OIDC window in Appsmith’s Admin Settings (link to that section to remind the user how to get this).
```

> ## Configure AWS Cognito with Appsmith

Reviewer: ```suggestion
## Configure Appsmith to use Amazon Cognito
```

> On your Appsmith instance, go to *Admin Settings> Authentication> OIDC*, click Enable and follow the steps below:

Reviewer: this line and point no 1 should be above **Create user pool** section. You can name the section 

**Enable OIDC on Appsmith**

1. Copy the redirect URL and add it as a Callback URL under the Hosted UI section of your AWS Cognito User pool.

![](/img/AWS-cognito_callback-url.png)

> 

Reviewer: Repeat step 1 for this section

> 2. Add the client ID and client secret key copied from the App Client in your user pool to the respective input boxes.

Reviewer: if we have a table for the rest of the input boxes, why not add this also there

> 3. Add the remaining configurations in the format shown below:

Reviewer: ```suggestion
3. Refer to the table below when entering data in the input boxes to configure the instance:
```

>     | OIDC configuration field in Appsmith | AWS Cognito URL |

Reviewer: ```suggestion
    | Configuration field | Description |
```

Author: It's not the description. It tells the users how to generate the URL from the information you get from Cognito.

    | -------------------------------------|----------------------- |
    | Authorization URL                    | `https://<your_user-pool_domain>/oauth2/authorize`|
    | Token URL                            | `https://<your_user-pool_domain>/oauth2/token` |
    | User Info URL                        | `https://<your_user-pool_domain>/oauth2/userInfo` |
    | JWK Set URL                          | `https://cognito-idp.<Region>.amazonaws.com/<userPoolId>/.well-known/jwks.json` | 

>     **your_user-pool_domain** - The Cognito domain you created in the above step.

Reviewer: these should be in the description column itself

    **Region** - The region's code where your AWS Cognito is set up.

    **userPoolId** - The unique ID of the user pool you've created.

    ![](/img/AWS_Urls.png)

> ### Complete OIDC setup

Reviewer: no need of this separate section. The Save & Restart button can be part of the continuation of the above steps
